### ISSUE TITLE
False positive for invalid null pointer usage lint when creating raw slice pointer

### ISSUE THREAD

**Issue by ljgermain** [rust-lang/rust-clippy#13445]
### Summary

Null slice pointers are valid Rust (i.e. `*[T]`), and only a null slice reference (`&[T]`) is actually a problem and leads to undefined behavior, but Clippy unnecessarily complains that it is "invalid" to create a null slice pointer and falsely asserts that it must be non-null. The offending lines are here: https://github.com/rust-lang/rust-clippy/blob/dbcd82885f4730077fabd57340d7e297122bc596/clippy_lints/src/ptr.rs#L255-256

### Lint Name

invalid_null_ptr_usage

### Reproducer

Here is a reproducible example that erroneously emits the warning:

```rust
#[warn(clippy::invalid_null_ptr_usage)]

fn main() {
    let _: *const [u8] = std::ptr::slice_from_raw_parts(std::ptr::null(), 0);
}
```

The output is:

```
warning: pointer must be non-null
 --> src/main.rs:4:57
  |
4 |     let _: *const [u8] = std::ptr::slice_from_raw_parts(std::ptr::null(), 0);
  |                                                         ^^^^^^^^^^^^^^^^ help: change this to: `core::ptr::NonNull::dangling().as_ptr()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#invalid_null_ptr_usage
note: the lint level is defined here
 --> src/main.rs:1:8
  |
1 | #[warn(clippy::invalid_null_ptr_usage)]
```

### Version

rustc 1.83.0-nightly (7042c269c 2024-09-23)
binary: rustc
commit-hash: 7042c269c166191cd5d8daf0409890903df7af57
commit-date: 2024-09-23
host: x86_64-unknown-linux-gnu
release: 1.83.0-nightly
LLVM version: 19.1.0

**Comment by samueltardieu** [comment]
> Null slice pointers are valid Rust

I think Clippy is aligned with the [documentation of `std::slice::from_raw_parts()`](https://doc.rust-lang.org/stable/std/slice/fn.from_raw_parts.html) here: "`data` must be non-null and aligned even for zero-length slices".

**Comment by y21** [comment]
> > Null slice pointers are valid Rust
> 
> I think Clippy is aligned with the [documentation of `std::slice::from_raw_parts()`](https://doc.rust-lang.org/stable/std/slice/fn.from_raw_parts.html) here: "`data` must be non-null and aligned even for zero-length slices".

Note that this is about `std::ptr::slice_from_raw_parts`, not `std::slice::from_raw_parts`. The former is a safe function, so this usage can't be invalid/UB
